app/services/sso_service.rb

```diff
@@ -7,19 +7,22 @@ class SSOService
   include SentryLogging
   include ActiveModel::Validations
 
-  def initialize(saml_response)
-    raise 'SAML Response is required' if saml_response.nil?
-    @saml_response = saml_response
-    @saml_attributes = SAML::User.new(saml_response)
-    @existing_user = User.find(saml_attributes.user_attributes.uuid)
-    @new_user_identity = UserIdentity.new(saml_attributes.to_hash)
-    @new_user = init_new_user(new_user_identity, existing_user, saml_attributes.changing_multifactor?)
-    @new_session = Session.new(uuid: new_user.uuid)
+  def initialize(response)
+    raise 'SAML Response is not a OneLogin::RubySaml::Response' unless response.is_a?(OneLogin::RubySaml::Response)
+    @saml_response = response
+
+    if saml_response.is_valid?(true)
```

Comment:
I feel like this could use some refactoring for clarity, but not sure it's complex enough to warrant it yet. Seems like we could probably avoid branching within the initialization and lean more heavily on the activemodel validation layer?

Reply:
Yes, I think i want to break this all out further. Validate Saml response and handle persistence in a separate class. That separate class could be a form based object that validates all three models: User, Session, UserIdentity. Right now there isn't a clear separation of concerns. Once there is better separation of concerns, I think the AuthFailHandler could also go away and just live in SSOService.